memory/StackMap.h

```diff
@@ -0,0 +1,302 @@
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    size_t capacity() const { return _elements.size(); }
+    size_t size() const { return _count; }
+    bool empty() const { return _count == 0; }
+
+    const_iterator cbegin() const
+    {
+        uint32_t first = 0;
+        while (first != _end && !_elements[first].committed)
+        {
+            ++first;
+        }
+        return const_iterator(const_cast<ElementEntry*>(_elements.data()), first, _end);
+    }
+
+    const_iterator cend() const { return const_iterator(const_cast<ElementEntry*>(_elements.data()), _end, _end); }
+
+    const_iterator begin() const { return cbegin(); }
+    const_iterator end() const { return cend(); }
+    iterator begin() { return iterator(cbegin()); }
+    iterator end() { return iterator(_elements.data(), _end, _end); }
+
+    template <typename TypeName = T>
+    TypeName getItem(const KeyT& key, typename std::enable_if<std::is_pointer<TypeName>::value>::type* = nullptr)
+    {
+        auto it = find(key);
+        if (it != end())
+        {
+            return it->second;
+        }
+        return nullptr;
+    }
+
+    template <typename TypeName = T>
+    TypeName* getItem(const KeyT& key, typename std::enable_if<!std::is_pointer<TypeName>::value>::type* = nullptr)
+    {
+        auto it = find(key);
+        if (it != end())
+        {
+            return &it->second;
+        }
+        return nullptr;
+    }
+
+    template <typename TypeName = T>
+    const TypeName getItem(const KeyT& key,
+        typename std::enable_if<std::is_pointer<TypeName>::value>::type* = nullptr) const
+    {
+        return const_cast<StackMap<KeyT, T, SIZE>&>(*this).getItem(key);
+    }
+
+    template <typename TypeName = T>
+    const TypeName* getItem(const KeyT& key,
+        typename std::enable_if<!std::is_pointer<TypeName>::value>::type* = nullptr) const
+    {
+        return const_cast<StackMap<KeyT, T, SIZE>&>(*this).getItem(key);
+    }
+
+private:
+    uint32_t indexPosition(uint64_t hashValue, uint32_t offset) const { return (hashValue + offset) % _index.size(); }
+
+    void updateSpread(uint32_t i)
+    {
+        if (i >= _maxSpread)
+        {
+            _maxSpread = i + 1;
+        }
+    }
+
+    uint32_t findFreePosition()
+    {
+        for (size_t i = 0; i < _index.size(); ++i)
+        {
+            if (!_elements[_nextFreeEntry].committed)
+            {
+                const auto freePosition = _nextFreeEntry + 1;
+                _nextFreeEntry = (_nextFreeEntry + 1) % SIZE;
+                return freePosition;
+            }
+            _nextFreeEntry = (_nextFreeEntry + 1) % SIZE;
+        }
+
+        return 0;
+    }
+
+    std::array<IndexEntry, SIZE * 4> _index;
+    std::array<ElementEntry, SIZE> _elements;
+    const uint32_t _end;
+    uint32_t _nextFreeEntry = 0;
+    T _emptyObject;
+    uint32_t _maxSpread;
```

Comment:
`_maxSpread` on increase never, decrease. One option would be clean the `spread` on `clean` but I guess call `clean` function will be rare cases.

Reply:
thanks, it should be reset.